File: model/src/main/scala/ru/mipt/acsl/decode/model/CommandOrTmMessage.scala
package ru.mipt.acsl.decode.model

import ru.mipt.acsl.decode.model.component.Command
import ru.mipt.acsl.decode.model.component.message.TmMessage

/**
  * @author Artem Shein
  */
trait CommandOrTmMessage extends Referenceable {

  def toEither: Either[Command, TmMessage]

}

object CommandOrTmMessage {

  private final case class CommandOrTmMessageImpl(toEither: Either[Command, TmMessage]) extends CommandOrTmMessage

  def apply(either: Either[Command, TmMessage]): CommandOrTmMessage = CommandOrTmMessageImpl(either)

}

File: model/src/main/java/ru/mipt/acsl/decode/model/CommandOrTmMessage.java
package ru.mipt.acsl.decode.model;

import ru.mipt.acsl.decode.model.component.Command;
import ru.mipt.acsl.decode.model.component.message.TmMessage;
import ru.mipt.acsl.decode.model.naming.Container;

import java.util.Optional;

/**
 * Created by metadeus on 06.06.16.
 */
public interface CommandOrTmMessage extends Container {

    static CommandOrTmMessage newInstance(Command command) {
        return new CommandOrTmMessageImpl(command);
    }

    static CommandOrTmMessage newInstance(TmMessage tmMessage) {
        return new CommandOrTmMessageImpl(tmMessage);
    }

    boolean isCommand();

    boolean isTmMessage();

    Optional<Command> command();

    Optional<TmMessage> tmMessage();

    @Override
    default <T> T accept(ContainerVisitor<T> visitor) {
        return visitor.visit(this);
    }
}
